sdks/java/io/hadoop-input-format/src/test/java/org/apache/beam/sdk/io/hadoop/inputformat/unit/tests/HadoopInputFormatIOTest.java

```diff
@@ -0,0 +1,796 @@
+   * InputSplit list having some null values.
+   */
+  @Test
+  public void testSplitIntoBundlesIfGetSplitsReturnsListHavingNullValues() throws Exception {
+    HadoopInputFormatBoundedSource<Text, Employee> hifSource = getTestHIFSource(
+            BadGetSplitsInputFormat.class,
+            Text.class,
+            Employee.class,
+            WritableCoder.of(Text.class),
+            AvroCoder.of(Employee.class));
+    thrown.expect(IOException.class);
+    thrown.expectMessage(HadoopInputFormatIOContants.COMPUTESPLITS_NULL_SPLIT_ERROR_MSG);
+    hifSource.getEstimatedSizeBytes(p.getOptions());
+    thrown.expect(IOException.class);
+    thrown.expectMessage(HadoopInputFormatIOContants.COMPUTESPLITS_NULL_SPLIT_ERROR_MSG);
+    hifSource.splitIntoBundles(0, p.getOptions());
+  }
+
+  /**
+   * This test validates functionality of {@link HadoopInputFormatIO} if user sets wrong key class
+   * and value class.
+   */
+  @Test
+  public void testHIFSourceIfUserSetsWrongKeyOrValueClass() throws Exception {
+    List<BoundedSource<KV<String, String>>> boundedSourceList = getBoundedSourceList(
+            NewObjectsEmployeeInputFormat.class,
+            String.class,
+            String.class,
+            StringUtf8Coder.of(),
+            StringUtf8Coder.of());
+    List<KV<String, String>> bundleRecords = new ArrayList<>();
+    for (BoundedSource<KV<String, String>> source : boundedSourceList) {
+      thrown.expect(ClassCastException.class);
+      List<KV<String, String>> elems = SourceTestUtils.readFromSource(source, p.getOptions());
+      bundleRecords.addAll(elems);
+    }
+  }
+
+  /**
+   * This test validates records emitted in PCollection are immutable if InputFormat's recordReader
+   * returns same objects(i.e. same locations in memory) but with updated values for each record.
+   */
+  @Test
+  public void testImmutablityOfOutputOfReadIfRecordReaderObjectsAreMutable() throws Exception {
+    List<BoundedSource<KV<Text, Employee>>> boundedSourceList = getBoundedSourceList(
+                ReuseObjectsEmployeeInputFormat.class,
+                Text.class,
+                Employee.class,
+                WritableCoder.of(Text.class),
+                AvroCoder.of(Employee.class));
+            List<KV<Text, Employee>> bundleRecords = new ArrayList<>();
+    for (BoundedSource<KV<Text, Employee>> source : boundedSourceList) {
+      List<KV<Text, Employee>> elems = SourceTestUtils.readFromSource(source, p.getOptions());
+      bundleRecords.addAll(elems);
+    }
+    List<KV<Text, Employee>> referenceRecords = TestEmployeeDataSet.getEmployeeData();
+    assertThat(bundleRecords, containsInAnyOrder(referenceRecords.toArray()));
+  }
+
+  /**
+   * Test reading if InputFormat implements {@link org.apache.hadoop.conf.Configurable
+   * Configurable}.
+   */
+  @Test
+  public void testReadingWithConfigurableInputFormat() throws Exception {
```

Comment:
Added a Boolean, Need to discuss this further  "I would suggest adding an explicit bool on CEIF for "was setConf called?" and then asserting on that in this test".
 Lets discuss in call today.